File: matlab/h5_reader.m
% Returns ONE patch as HxWxC (single) from So2Sat LCZ42 H5 files.
%
% Matteo Rambaldi — Thesis utilities

function X = h5_reader(pth, index, modality)

    arguments
        pth (1,1) string
        index (1,1) double {mustBeInteger, mustBePositive}
        modality (1,1) string
    end

    modality = upper(modality);
    switch modality
        case "MS"
            dataset = '/sen2';
        case "SAR"
            dataset = '/sen1';
        otherwise
            error("Unknown modality: %s", modality);
    end

    info = h5info(pth, dataset);
    dims = info.Dataspace.Size;  % [C, H, W, N]

    C = dims(1);
    H = dims(2);
    W = dims(3);

    % Dynamically read the correct shape for 32x32 or 64x64
    X4 = h5read(pth, dataset, [1 1 1 index], [C H W 1]);
    X = permute(X4, [2 3 1 4]);   % [H W C 1] → [H W C]
    X = single(X(:,:, :, 1));
end

File: fusion_ensembles_no_SR/scripts/h5_reader.py
import h5py
import numpy as np

def h5_reader(path: str, index: int, modality: str) -> np.ndarray:
    modality = modality.upper()
    if modality == "MS":
        dataset = "/sen2"
    elif modality == "SAR":
        dataset = "/sen1"
    else:
        raise ValueError(f"Unknown modality: {modality}")

    with h5py.File(path, "r") as f:
        dset = f[dataset]
        # h5py uses (N, H, W, C)
        if index < 0 or index >= dset.shape[0]:
            raise IndexError(f"Index {index} out of range 0–{dset.shape[0]-1}")

        patch = dset[index]  # shape: (H, W, C)
        patch = np.array(patch, dtype=np.float32)
        return patch